src/gui/guiButtonImage.cpp

```diff
@@ -30,109 +30,48 @@ using namespace irr;
 {
 	m_image = Environment->addImage(
 			core::rect<s32>(0,0,rectangle.getWidth(),rectangle.getHeight()), this);
 	m_image->setScaleImage(isScalingImage());
 	sendToBack(m_image);
 }
 
-bool GUIButtonImage::OnEvent(const SEvent& event)
-{
-	bool result = GUIButton::OnEvent(event);
-
-	EGUI_BUTTON_IMAGE_STATE imageState = getImageState(isPressed(), m_foreground_images);
-	video::ITexture *texture = m_foreground_images[(u32)imageState].Texture;
-	if (texture != nullptr)
-	{
-		m_image->setImage(texture);
-	}
-
-	m_image->setVisible(texture != nullptr);
-
-	return result;
-}
-
-void GUIButtonImage::setForegroundImage(EGUI_BUTTON_IMAGE_STATE state,
-		video::ITexture *image, const core::rect<s32> &sourceRect)
+void GUIButtonImage::setForegroundImage(video::ITexture *image)
 {
-	if (state >= EGBIS_COUNT)
+	if (image == m_foreground_image)
 		return;
 
-	if (image)
+	if (image != nullptr)
 		image->grab();
 
-	u32 stateIdx = (u32)state;
-	if (m_foreground_images[stateIdx].Texture)
-		m_foreground_images[stateIdx].Texture->drop();
-
-	m_foreground_images[stateIdx].Texture = image;
-	m_foreground_images[stateIdx].SourceRect = sourceRect;
-
-	EGUI_BUTTON_IMAGE_STATE imageState = getImageState(isPressed(), m_foreground_images);
-	if (imageState == stateIdx)
-		m_image->setImage(image);
-}
-
-void GUIButtonImage::setForegroundImage(video::ITexture *image)
-{
-	setForegroundImage(gui::EGBIS_IMAGE_UP, image);
-}
-
-void GUIButtonImage::setForegroundImage(video::ITexture *image, const core::rect<s32> &pos)
-{
-	setForegroundImage(gui::EGBIS_IMAGE_UP, image, pos);
-}
-
-void GUIButtonImage::setPressedForegroundImage(video::ITexture *image)
-{
-	setForegroundImage(gui::EGBIS_IMAGE_DOWN, image);
-}
+	if (m_foreground_image != nullptr)
+		m_foreground_image->drop();
 
-void GUIButtonImage::setPressedForegroundImage(video::ITexture *image, const core::rect<s32> &pos)
-{
-	setForegroundImage(gui::EGBIS_IMAGE_DOWN, image, pos);
+	m_foreground_image = image;
+	m_image->setImage(image);
 }
 
-void GUIButtonImage::setHoveredForegroundImage(video::ITexture *image)
+//! Set element properties from a StyleSpec
+void GUIButtonImage::setFromStyle(const StyleSpec& style)
 {
-	setForegroundImage(gui::EGBIS_IMAGE_UP_MOUSEOVER, image);
-	setForegroundImage(gui::EGBIS_IMAGE_UP_FOCUSED_MOUSEOVER, image);
-}
-
-void GUIButtonImage::setHoveredForegroundImage(video::ITexture *image, const core::rect<s32> &pos)
-{
-	setForegroundImage(gui::EGBIS_IMAGE_UP_MOUSEOVER, image, pos);
-	setForegroundImage(gui::EGBIS_IMAGE_UP_FOCUSED_MOUSEOVER, image, pos);
-}
-
-void GUIButtonImage::setFromStyle(const StyleSpec &style, ISimpleTextureSource *tsrc)
-{
-	GUIButton::setFromStyle(style, tsrc);
+	GUIButton::setFromStyle(style);
 
 	video::IVideoDriver *driver = Environment->getVideoDriver();
 
 	const core::position2di buttonCenter(AbsoluteRect.getCenter());
 	core::position2d<s32> geom(buttonCenter);
 	if (style.isNotDefault(StyleSpec::FGIMG)) {
```

Comment:
Why is `FGIMG` not handled in `setFromStyle`, but `BGIMG` is?

Reply:
I'm not sure I understand, both are handled. GUIButtonImage is the only element that supports `FGIMG`, so the code for handling it is here. The base class implementation handles the shared style properties, like `BGIMG`.

Reply:
Oh, I missed such an obvious point. Thank you for explaining.